lldb/source/Core/ModuleList.cpp

```diff
@@ -755,11 +755,237 @@ size_t ModuleList::GetIndexForModule(const Module *module) const {
+    return result;
+  }
+
+  std::recursive_mutex &GetMutex() const { return m_list.GetMutex(); }
+
+private:
+  ModuleSP FindModuleInIndex(const Module *module_ptr) {
+    if (!module_ptr->GetFileSpec().GetFilename())
+      return ModuleSP();
+    ConstString name = module_ptr->GetFileSpec().GetFilename();
+    auto it = m_index.find(name);
+    if (it != m_index.end()) {
+      auto &vector = it->getSecond();
+      for (auto &module_sp : vector)
+        if (module_sp.get() == module_ptr)
+          return module_sp;
+    }
+    return ModuleSP();
+  }
+
+  bool FindModulesInIndex(const ModuleSpec &module_spec,
+                          ModuleList &matching_module_list) const {
+    auto it = m_index.find(module_spec.GetFileSpec().GetFilename());
+    if (it == m_index.end())
+      return false;
+    auto vector = it->getSecond();
+    bool found = false;
+    for (auto &module_sp : vector) {
+      if (module_sp->MatchesModuleSpec(module_spec)) {
+        matching_module_list.Append(module_sp);
+        found = true;
+      }
+    }
+    return found;
+  }
+
+  void AddToIndex(const ModuleSP &module_sp) {
+    auto name = module_sp->GetFileSpec().GetFilename();
+    if (name.IsEmpty())
+      return;
+    auto &vec = m_index[name];
+    vec.push_back(module_sp);
+  }
+
+  void RemoveFromIndex(const Module *module_ptr, bool if_orphaned = false) {
+    auto name = module_ptr->GetFileSpec().GetFilename();
+    auto it = m_index.find(name);
+    if (it == m_index.end())
+      return;
+    auto &vec = it->getSecond();
+    for (auto *it = vec.begin(); it != vec.end(); ++it) {
+      if (it->get() == module_ptr) {
+        // use_count == 2 means only held by index and list (orphaned)
+        if (!if_orphaned || it->use_count() == 2)
+          vec.erase(it);
+        break;
+      }
+    }
+  }
+
+  void ReplaceEquivalentInIndex(const ModuleSP &module_sp) {
+    RemoveEquivalentModulesFromIndex(module_sp);
+    AddToIndex(module_sp);
+  }
+
+  void RemoveEquivalentModulesFromIndex(const ModuleSP &module_sp) {
+    auto name = module_sp->GetFileSpec().GetFilename();
+    if (name.IsEmpty())
+      return;
+
+    auto it = m_index.find(name);
+    if (it == m_index.end())
+      return;
+
+    // First remove any equivalent modules. Equivalent modules are modules
+    // whose path, platform path and architecture match.
+    ModuleSpec equivalent_module_spec(module_sp->GetFileSpec(),
+                                      module_sp->GetArchitecture());
+    equivalent_module_spec.GetPlatformFileSpec() =
+        module_sp->GetPlatformFileSpec();
+
+    auto &vec = it->getSecond();
+    // Iterate backwards to minimize element shifting during removal
+    for (int i = vec.size() - 1; i >= 0; --i) {
+      auto *it = vec.begin() + i;
+      if ((*it)->MatchesModuleSpec(equivalent_module_spec))
+        vec.erase(it);
+    }
+  }
+
+  /// Remove orphans from both the index and list, if orphaned.
```

Comment:
I'm not sure I understand this sentence. Could you elaborate here? It sounds circular: "remove orphans if orphans"

Reply:
Yeah not a great comment, I'm updating it.